src/bookshelf/src/bksw_wm_sql_object.erl

```diff
@@ -0,0 +1,341 @@
+            Error
+    end.
+
+fully_read(#context{} = Ctx, []) ->
+    fully_read(send_streamed_body(Ctx), []);
+fully_read({_, done}, Accum) ->
+    lists:reverse(Accum);
+fully_read({error, Error}, Accum) ->
+    error_logger:error_msg("Error occurred during content download: ~p~n", [Error]),
+    lists:reverse(Accum);
+fully_read({Data, Next}, Accum) ->
+    fully_read(Next(), [Data | Accum]).
+
+%% Creates or replaces a file.
+%% Existing files have their data segment replaced.
+%% New files will have no file_id, and neeed a new file name entry
+finalize_maybe_create_file(_Rq, _Ctx,
+                           #db_file{file_id = undefined,
+                                    bucket_name = BucketName,
+                                    name = Name,
+                                    data_id = DataId}) ->
+    case bksw_sql:create_file_link_data(BucketName, Name, DataId) of
+        {ok, _} ->
+            ok;
+        {error, Reason} ->
+            error_logger:error_msg("create_file_link_data ~p~n", [Reason]),
+            fail
+    end;
+%% If we have a file_id, we're replacing an old file
+finalize_maybe_create_file(_Rq, _Ctx,
+                           #db_file{file_id = FileId,
+                                    data_id = DataId}) ->
+    case bksw_sql:link_file_data(FileId, DataId) of
+        ok ->
+            ok;
+        {error, Reason} ->
+            error_logger:error_msg("link_file_data ~p~n", [Reason]),
+            fail
+    end.
+
+%%
+%% Takes a chunk from webmachine, and streams it out to the db
+%%
+%% A future version of this may want to re-chunk things;
+write_streamed_body({Data, done}, Rq0,
+                    #context{entry_md = #db_file{data_id = DataId} = File0,
+                             transfer_state = #file_transfer_state{next_chunk = ChunkId} = TransferState0} = Ctx0 ) ->
+    #file_transfer_state{next_chunk = ChunkCount} = TransferState1 =
+        case Data of
+            <<"">> ->
+                TransferState0;
+            Data ->
+                ok = bksw_sql:add_file_chunk(DataId, ChunkId, Data),
+                bksw_sql:update_transfer_state(TransferState0, Data, 1)
+        end,
+
+
+    #db_file{data_size = Size,
+             hash_md5 = HashMd5,
+             hash_sha256 = HashSha256,
+             hash_sha512 = HashSha512} = File1 =
+        bksw_sql:finalize_transfer_state(TransferState1, File0),
+    Ctx1 = Ctx0#context{entry_md = File1},
+    bksw_sql:update_metadata(DataId, Size, ChunkCount, HashMd5, HashSha256, HashSha512),
+
+    case get_header('Content-MD5', Rq0) of
+        undefined ->
+            Rq1 = bksw_req:with_etag(base64:encode(HashMd5), Rq0),
+            %% Failures here turn into 500s and stack traces; we might want to do better if this
+            %% happens much.
+            ok = finalize_maybe_create_file(Rq1, Ctx0, File1),
+            {true, wrq:set_response_code(202, Rq1), Ctx1};
```

Comment:
As far as I can tell, this, and the error cases below, will all return HTTP 204 because we aren't sending a response body and webmachine doesn't seem to care what we've set in the second argument. If we want to set a response code I think we need to use `{{halt, 202}, Rq1, Ctx1}` as our return value here.  However, I'll note that this same pattern is used in a number of places so I'm guessing most of our client code doesn't care too much about 202 vs 204; however, we definitely want to make sure we return non-2XX at line 328.